pkg/kubelet/status/generate.go

```diff
@@ -300,3 +318,46 @@ func generatePodReadyConditionForTerminalPhase(pod *v1.Pod, oldPodStatus *v1.Pod
 
 	return condition
 }
+
+func GenerateAllContainersRestartingCondition(pod *v1.Pod, podStatus *kubecontainer.PodStatus, oldPodStatus *v1.PodStatus, podPhase v1.PodPhase) v1.PodCondition {
+	if podPhase == v1.PodSucceeded {
+		return v1.PodCondition{
+			Type:   v1.AllContainersRestarting,
+			Status: v1.ConditionFalse,
+			Reason: PodCompleted,
+		}
+	}
+	if podPhase == v1.PodFailed {
+		return v1.PodCondition{
+			Type:   v1.AllContainersRestarting,
+			Status: v1.ConditionFalse,
+			Reason: PodFailed,
+		}
+	}
+
+	if kubecontainer.AllContainersRestartCleanedUp(pod, podStatus) {
+		return v1.PodCondition{
+			Type:   v1.AllContainersRestarting,
+			Status: v1.ConditionFalse,
+		}
+	}
+	if kubecontainer.ShouldAllContainersRestart(pod, podStatus, oldPodStatus) {
+		return v1.PodCondition{
+			Type:    v1.AllContainersRestarting,
+			Status:  v1.ConditionTrue,
+			Reason:  RestartAllContainersStarted,
+			Message: "container exited with restart policy rule",
+		}
+	}
+
+	// Keep the old condition if the restart is not triggered or cleanup has not finished.
+	for _, cond := range oldPodStatus.Conditions {
+		if cond.Type == v1.AllContainersRestarting {
+			return cond
+		}
+	}
+	return v1.PodCondition{
```

Comment:
Do you really want this condition to be present on every pod? At the very least, it seems like we should omit it on pods that don't have any RestartAllContainers restart rules.